074-connector-offsets-support.md

```diff
@@ -0,0 +1,226 @@
+This matches the name used in the list operation, so that a user can list offsets, then edit the ConfigMap the operator created, then request the offsets be altered using that same ConfigMap.
+
+The KafkaConnector CRD will also be updated to contain a new property called `alterOffsetsCM` that when set will override the name of the ConfigMap the operator uses to populate the request body.
+
+Notes:
+* The format of the request body will be validated to make sure it is valid JSON, but the operator won't perform any further validation, instead relying on the API endpoint to return a reasonable error message.
+* If the request fails the KafkaConnector CR status will be updated to include a warning message with the response from the API, e.g.
+  ```yaml
+  - lastTransitionTime: "2024-06-04T08:44:15.913138115Z"
+    message: Failed to alter the connector offsets due to "message from endpoint".
+    reason: AlterOffsets
+    status: "True"
+    type: Warning
+  ```
+* Strimzi will not attempt to verify that the connector is already stopped, it will rely on the API failing if that is the case.
+  This is to simplify the logic and means we don't have to account for what happens if the KafkaConnector spec has been updated but the connector isn't stopped yet.
+
+### Resetting offsets
+
+To reset the offsets for a particular connector the user will annotate the KafkaConnector resource with the new annotation `strimzi.io/connectorOffsets` set to `reset`.
+When the operator observes this annotation it will use the DELETE /connectors/{connector}/offsets endpoint to reset all offsets for the connector.
+When the reset has been completed the operator will remove this annotation from the resource.
+
+If the reset fails the operator will add a condition to the KafkaConnector resource called `ResetOffsets` to communicate the failure and remove the annotation.
+
+Strimzi will shortcut and automatically fail to do the reset if the KafkaConnector resource is not stopped.
+The user can update the KafkaConnector to stop the connector and reset the offsets at the same time.
+In that case the operator will first stop the connector, and once that API call returns, then it will make the call to reset the offsets.
+
+#### Resetting on delete
+
+It is possible for the operator to help the user in cleaning up resources.
+The KafkaConnector CR will have a new property added called `resetOffsetsOnDelete`.
+This will be optional and when not present it will default to `false`.
+When `resetOffsetsOnDelete` is set to `true` the operator will perform the following steps when it sees a connector needs to be deleted:
+1. Stop the connector using the PUT /connectors/{connector}/stop endpoint 
+2. Reset offsets using the DELETE /connectors/{connector}/offsets endpoint 
+3. Delete the connector the DELETE /connectors/{connector} endpoint
+
+Notes:
+* Steps 1 & 2 are new compared to the existing steps the operator takes when deleting connectors
```

Comment:
Can we really guarantee this? What if the Connect is not working anymore or is deleted together with the connector etc.?

Reply:
I think we can give the same level of guarantees as we do with deleting the connector.

For example if Connect isn't working or has been deleted the call to the delete the connector would also fail, so if in future a new Connect cluster was created that re-used the same topics the connector would be started on that Connect cluster automatically, since it wouldn't have been removed from the Connect configs topic.

Reply:
I'm not sure this is so simple:
* The connector will stay in the Connect topics and be recovered -> but deleted right away if the KafkaConnector CR is missing. So you can probably call it a _very eventual consistency_ :-).
* In your case, the connector will be gone. But the offset will not be reset and later when the Connect cluster is started again, you will not know you were supposed to delete some offsets.

Reply:
In the proposal I'm suggesting that when we delete the connector we first stop the connector and reset the offsets, so effectively adding those two steps before this line https://github.com/strimzi/strimzi-kafka-operator/blob/main/cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectAssemblyOperator.java#L488 I'm not following why that would result in the connector being deleted before we can reset the offsets.

However, I have spotted a different problem, which is since we only know to delete the connector because the KafkaConnector CR is missing, we can't use a property on the KafkaConnector CR to tell us whether to reset the offsets first can we 🤔 I think it only works if we always reset the offsets, which as previously discussed we probably don't want

Reply:
> In the proposal I'm suggesting that when we delete the connector we first stop the connector and reset the offsets, so effectively adding those two steps before this line https://github.com/strimzi/strimzi-kafka-operator/blob/main/cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectAssemblyOperator.java#L488 I'm not following why that would result in the connector being deleted before we can reset the offsets.

This is the normal flow. Sure, you can do it there. But in many of the edge cases, you will not.

> However, I have spotted a different problem, which is since we only know to delete the connector because the KafkaConnector CR is missing, we can't use a property on the KafkaConnector CR to tell us whether to reset the offsets first can we 🤔 I think it only works if we always reset the offsets, which as previously discussed we probably don't want

If I get right what you mean, then check the `connectorEventHandler` in KafkaConnectAssemblyOperator. You should (IIRC) get there the event handled inclusing the deleted CR. So you can see what the settings were. But of course, if the operator misses the watch because it was for example rolling at that time, it will not have it. 

That is the point I'm making here. You cannot easily make it 100% reliable. So is it worth it with the caveat it might not work? Or is it better to leave it to users to do it as combination of steps and take the respnsibility?